tdms/tests/unit/test_hdf5_io.cpp

```diff
@@ -44,29 +56,151 @@ TEST_CASE("Test file I/O construction/destruction.") {
     CHECK_THROWS(HDF5Reader(tmp.string() + "/this_file_doesnt_exist.h5"));
   }
 
+  SECTION("Check can't read nonexistent data.") {
+    {
+      HDF5Writer fw(tmp.string() + "/this_file_does_exist_but_is_empty.h5");
+      CHECK(fw.is_ok());
+
+    }// Destructor called as we leave scope.
+
+    double data[1];
+    HDF5Reader fr(tmp.string() + "/this_file_does_exist_but_is_empty.h5");
+    CHECK_THROWS(fr.read("nonexistantdata", data));
```

Comment:
Not the end of the world, but we're actually not sure which exception is being thrown here. Presumably there'd also be a HDF5 exception if `data[1]` was too small or the wrong type or whatever.

Reply:
HDF5 (un)helpfully wraps all it's errors into HDF5Errors which aren't particularly readable. In this case I think it's their equivalent of a "file not found" error but it's not actually called that :roll_eyes:

Reply:
Ah, I just mean that we're expecting some kind of specific data exception or something. H5 seems to have a [pretty nice set of exceptions](https://docs.hdfgroup.org/archive/support/HDF5/doc1.8/cpplus_RM/class_h5_1_1_exception.html).

"File not found" is probably a `H5::FileIException`. This non-existant data is _probably_ something like a `H5::DataSetException`. So we should specify the thing we expect to be thrown is thrown. Not just any general exception.

Reply:
Ugh. They're both `H5::FileIException`s. Which I guess makes sense.... trying to read something from *a file* that isn't there.

Reply:
Catch2 has a [`{REQUIRE,CHECK}_THROWS_WITH`](https://github.com/catchorg/Catch2/blob/devel/docs/assertions.md#exceptions) if you want to be super-specific and check the exception message to distinguish the cases?

Reply:
I went for `CHECK_THROWS_AS` but not `WITH` (because the exception string is quite long).